### DISCUSSION TITLE
Nested / Anonymous (auxiliary) recursive functions

### DISCUSSION THREAD

**Discussion by nohehf** [gleam-lang/gleam#3957] [General]
In functional programming, auxiliary recursive functions are a common pattern I think (at least from my experience).
For instance in ocaml you can implement factorial as:

```ocaml
let factorial n =
  let rec aux acc n =
    if n = 0 then acc
    else aux (acc * n) (n - 1)
  in aux 1 n
```
*(I know that this is not the best factorial implementation, and that in this case the aux function can be considered and anti-pattern, but that is an other debate, and the example is simple).*

This does not seem to be possible in gleam however as in gleam only module-level functions can be named:
```gleam 
pub fn factorial(n) {
  fn aux() {} //  Syntax error I was not expecting this. Found a name, expected one of:  - `(` Hint: Only module-level functions can be named.
}
```

And anonymous function handle (variable) won't be in scope of the said function

```gleam
pub fn factorial(n) {
  let aux = fn() {
     aux() //  Unknown variable The name `aux` is not in scope here. (hence the rec keyword in ocaml)
  } 
}
```

I tried to pass the function as a callback argument as a ungly workaround, but cannot be typed correctly I think (recursive types):
```gleam
// invalid tentative
pub fn factorial(n) {
  // aux cannot be typed (as it would result in fn(Int, Int, fn(Int, Int, ...))
  let aux = fn(acc: Int, n: Int, aux) {
    case n {
      0 -> acc
      _ -> aux(acc * n, n - 1, aux)
    }
  }
  aux(1, n, aux)
}
```

So I wonder:
- What is the rationale behind the rule "only module-level functions can be named" ?
- Is there another way to achieve this ? (Other than two module-level functions) ?
- Would a `rec` like keyword be envisageable in gleam ?

`rec` would allow something like:
```gleam
// invalid tentative
pub fn factorial(n) {
  let rec aux = fn(acc: Int, n: Int) {
    case n {
      0 -> acc
      _ -> aux(acc * n, n - 1)
    }
  }
  aux(1, n)
}
```
(or maybe rec would not even be needed ? That would be even cooler)

Thanks!

PS: On a personnal note, and It seems that this have been discussed in https://github.com/gleam-lang/gleam/issues/346 and https://github.com/gleam-lang/gleam/issues/349, but I think having some sort of let rec in, or another solution to make top level and nested function the same would be really nice. I've been loving gleam so far, but not having a single type of function (top level could be `let pub top_level_function = fn () {}` feels a bit weird to me. The ability to do the same things in or out of top level context would be awesome. Thanks again for the incredible work tho, gleam has been a joy :))

**Comment by lpil** [comment]
We won't be adding recursive let as you can use a top level function of the fix point combinator function to the same effect.

**Reply by nohehf** [reply]
What do you mean by 
> of the fix point combinator function 

And don't you think it could be nice to have the same behavior from top level and non- top level functions ?

Thanks for the quick answer !

**Reply by lpil** [reply]
It's a function that can make other functions recursive https://hexdocs.pm/funtil/funtil.html#fix

> And don't you think it could be nice to have the same behavior from top level and non- top level functions ?

No, this is by design.

**Reply by nohehf** [reply]
> It's a function that can make other functions recursive https://hexdocs.pm/funtil/funtil.html#fix

Thanks that's exactly what I was looking for

**Comment by Olian04** [comment]
@nohehf I've written a library for this exact usecase. It implements the fixedpoint combinator, which let's you create recursive blocks in Gleam with one line of code. https://github.com/Olian04/gleam_recursive

**Reply by lpil** [reply]
See also the funtil package!

**Reply by Olian04** [reply]
@lpil I did concider that library when I wrote mine. But I didn't trust that library since the link (on hex) to the repository points to the Lustre repo. This makes it a lot harder to verify the contents of the package before downloading.

**Reply by lpil** [reply]
You should never use the repository link to verify packages. It is unsafe. There is no guarantee that what is in the git repo is what is on Hex.

You must always verify the code on Hex itself. https://preview.hex.pm/preview/funtil/1.0.0/show/src/funtil.gleam

**Reply by Olian04** [reply]
I agree @lpil, I'm just giving my reasoning as to why I didn't trust the author.
